File: app/src/main/java/com/innovation4you/napking/data/NapKingRepository.java
package com.innovation4you.napking.data;

import com.google.android.gms.maps.model.LatLng;
import com.innovation4you.napking.data.provider.INapKingDataProvider;
import com.innovation4you.napking.model.RestStop;
import com.innovation4you.napking.model.SearchResult;

import java.util.List;

public class NapKingRepository implements INapKingDataProvider {

	private static NapKingRepository instance;

	private INapKingDataProvider dataProvider;

	public static void init(INapKingDataProvider dataProvider) {
		if (instance == null) {
			instance = new NapKingRepository(dataProvider);
		}
	}

	public static NapKingRepository get() {
		return instance;
	}

	private NapKingRepository(INapKingDataProvider dataProvider) {
		this.dataProvider = dataProvider;
	}


	@Override
	public List<RestStop> findRestStops(String condition) {
		return dataProvider.findRestStops(condition);
	}

	@Override
	public List<SearchResult> search(LatLng source, LatLng destination, int minutesLeft) {
		return dataProvider.search(source, destination, minutesLeft);
	}

	@Override
	public RestStop findRestStopById(long id) {
		return dataProvider.findRestStopById(id);
	}
}

File: app/src/main/java/com/innovation4you/napking/data/NapKingRepository.kt
package com.innovation4you.napking.data

import com.google.android.gms.maps.model.LatLng
import com.innovation4you.napking.data.provider.INapKingDataProvider
import com.innovation4you.napking.model.RestStop
import com.innovation4you.napking.model.SearchResult

class NapKingRepository private constructor(private val dataProvider: INapKingDataProvider) : INapKingDataProvider {


    override fun findRestStops(condition: String): List<RestStop> {
        return dataProvider.findRestStops(condition)
    }

    override fun search(source: LatLng, destination: LatLng, minutesLeft: Int): List<SearchResult> {
        return dataProvider.search(source, destination, minutesLeft)
    }

    override fun findRestStopById(id: Long): RestStop? {
        return dataProvider.findRestStopById(id)
    }

    companion object Factory {

        private lateinit var instance: NapKingRepository;

        fun init(dataProvider: INapKingDataProvider) {
            instance = NapKingRepository(dataProvider)
        }

        fun get(): NapKingRepository {
            return instance
        }
    }
}
